File: app/src/main/java/com/peterlaurence/trekme/ui/mapimport/MapArchiveViewHolder.java
package com.peterlaurence.trekme.ui.mapimport;

import android.graphics.Color;
import android.graphics.PorterDuff;
import android.view.View;
import android.view.ViewStub;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.peterlaurence.trekme.R;
import com.peterlaurence.trekme.core.map.mapimporter.MapImporter;

/**
 * @author peterLaurence on 22/12/17.
 */
public class MapArchiveViewHolder extends RecyclerView.ViewHolder {
    int position;
    ConstraintLayout layout;
    TextView mapArchiveName;

    /* The indeterminate unzip progressBar and its stub */
    private ViewStub stubProgressBarUnzip;
    private ProgressBar progressBarIndUnzip;

    /* Those view below could also be loaded later using ViewStub */
    private ProgressBar progressBarHorizontal;
    private ImageView iconMapExtracted;
    private ImageView iconMapExtractionError;
    private TextView extractionLabel;
    private ProgressBar progressBarIndMapCreation;
    private ImageView iconMapCreated;
    private TextView mapCreationLabel;


    public MapArchiveViewHolder(View itemView) {
        super(itemView);
        layout = itemView.findViewById(R.id.map_archive_contraint_layout);
        mapArchiveName = itemView.findViewById(R.id.map_archive_name);
        progressBarHorizontal = itemView.findViewById(R.id.unzip_progressbar);
        progressBarHorizontal.setMax(100);
        stubProgressBarUnzip = itemView.findViewById(R.id.stub_extraction_ind_progressbar);
        iconMapExtracted = itemView.findViewById(R.id.extraction_done);
        iconMapExtractionError = itemView.findViewById(R.id.extraction_error);
        extractionLabel = itemView.findViewById(R.id.extraction_txtview);
        progressBarIndMapCreation = itemView.findViewById(R.id.mapcreation_ind_progressbar);
        progressBarIndMapCreation.getIndeterminateDrawable().setColorFilter(Color.GRAY, PorterDuff.Mode.MULTIPLY);
        iconMapCreated = itemView.findViewById(R.id.mapcreation_done);
        mapCreationLabel = itemView.findViewById(R.id.mapcreation_txtview);
    }

    public void init() {
        progressBarHorizontal.setVisibility(View.GONE);
        if (progressBarIndUnzip != null) {
            progressBarIndUnzip.setVisibility(View.GONE);
        }
        iconMapExtracted.setVisibility(View.GONE);
        iconMapExtractionError.setVisibility(View.GONE);
        extractionLabel.setVisibility(View.GONE);
        progressBarIndMapCreation.setVisibility(View.GONE);
        iconMapCreated.setVisibility(View.GONE);
        mapCreationLabel.setVisibility(View.GONE);
    }

    /**
     * Init views based on view stubs.
     */
    private void initProgressBar() {
        if (progressBarIndUnzip == null) {
            progressBarIndUnzip = (ProgressBar) stubProgressBarUnzip.inflate();
            progressBarIndUnzip.getIndeterminateDrawable().setColorFilter(Color.GRAY, PorterDuff.Mode.MULTIPLY);
        }
    }

    public void onProgress(int p) {
        initProgressBar();
        progressBarHorizontal.setVisibility(View.VISIBLE);
        progressBarIndUnzip.setVisibility(View.VISIBLE);
        extractionLabel.setVisibility(View.VISIBLE);
        progressBarHorizontal.setProgress(p);
    }

    public void onUnzipFinished() {
        progressBarHorizontal.setVisibility(View.GONE);
        progressBarIndUnzip.setVisibility(View.GONE);
        extractionLabel.setVisibility(View.VISIBLE);
        mapCreationLabel.setVisibility(View.VISIBLE);
        iconMapExtracted.setVisibility(View.VISIBLE);
        progressBarIndMapCreation.setVisibility(View.VISIBLE);
    }

    public void onUnzipError() {
        initProgressBar();
        progressBarIndUnzip.setVisibility(View.GONE);
        iconMapExtractionError.setVisibility(View.VISIBLE);
        extractionLabel.setText(R.string.extraction_error);
    }

    public void onMapImported(MapImporter.MapParserStatus status) {
        switch (status) {
            case EXISTING_MAP:
                mapCreationLabel.setText(R.string.imported_untouched);
                break;
            case UNKNOWN_MAP_ORIGIN:
            case NO_MAP:
                mapCreationLabel.setText(R.string.map_import_error);
                break;
        }
        progressBarIndMapCreation.setVisibility(View.GONE);
        iconMapCreated.setVisibility(View.VISIBLE);
        mapCreationLabel.setVisibility(View.VISIBLE);
    }
}

File: app/src/main/java/com/peterlaurence/trekme/features/mapimport/presentation/ui/MapArchiveViewHolder.kt
package com.peterlaurence.trekme.features.mapimport.presentation.ui

import android.graphics.Color
import android.graphics.ColorFilter
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.recyclerview.widget.RecyclerView
import com.peterlaurence.trekme.R
import com.peterlaurence.trekme.core.map.domain.models.MapParseStatus
import com.peterlaurence.trekme.databinding.MapArchiveCardBinding

/**
 * @author P.Laurence on 22/12/17 -- Converted to Kotlin on 02/10/20
 */
class MapArchiveViewHolder(binding: MapArchiveCardBinding) : RecyclerView.ViewHolder(binding.root) {
    var layout: ConstraintLayout = binding.root
        private set
    var mapArchiveName: TextView = binding.mapArchiveName
        private set

    private val progressBarHorizontal: ProgressBar = binding.unzipProgressbar
    private val iconMapExtracted: ImageView = binding.extractionDone
    private val iconMapExtractionError: ImageView = binding.extractionError
    private val extractionLabel: TextView = binding.extractionTxtview
    private val progressBarIndUnzip: ProgressBar = binding.extractionIndProgressbar
    private val progressBarIndMapCreation: ProgressBar = binding.mapcreationIndProgressbar
    private val iconMapCreated: ImageView = binding.mapcreationDone
    private val mapCreationLabel: TextView = binding.mapcreationTxtview

    private val colorFilter: ColorFilter? by lazy {
        BlendModeColorFilterCompat.createBlendModeColorFilterCompat(Color.GRAY, BlendModeCompat.MODULATE)
    }

    fun init() {
        progressBarHorizontal.visibility = View.GONE
        progressBarIndUnzip.visibility = View.GONE
        iconMapExtracted.visibility = View.GONE
        iconMapExtractionError.visibility = View.GONE
        extractionLabel.visibility = View.GONE
        progressBarIndMapCreation.visibility = View.GONE
        iconMapCreated.visibility = View.GONE
        mapCreationLabel.visibility = View.GONE
    }

    fun onProgress(p: Int) {
        progressBarHorizontal.visibility = View.VISIBLE
        progressBarIndUnzip.visibility = View.VISIBLE
        extractionLabel.visibility = View.VISIBLE
        progressBarHorizontal.progress = p
    }

    fun onUnzipFinished() {
        progressBarHorizontal.visibility = View.GONE
        progressBarIndUnzip.visibility = View.GONE
        extractionLabel.visibility = View.VISIBLE
        mapCreationLabel.visibility = View.VISIBLE
        iconMapExtracted.visibility = View.VISIBLE
        progressBarIndMapCreation.visibility = View.VISIBLE
    }

    fun onUnzipError() {
        progressBarIndUnzip.visibility = View.GONE
        iconMapExtractionError.visibility = View.VISIBLE
        extractionLabel.setText(R.string.extraction_error)
    }

    fun onMapImported(status: MapParseStatus) {
        when (status) {
            MapParseStatus.NEW_MAP -> mapCreationLabel.setText(R.string.imported_new_map)
            MapParseStatus.EXISTING_MAP -> mapCreationLabel.setText(R.string.imported_untouched)
            MapParseStatus.UNKNOWN_MAP_ORIGIN, MapParseStatus.NO_MAP -> mapCreationLabel.setText(R.string.map_import_error)
        }
        progressBarIndMapCreation.visibility = View.GONE
        iconMapCreated.visibility = View.VISIBLE
        mapCreationLabel.visibility = View.VISIBLE
    }

    init {
        progressBarHorizontal.max = 100
        progressBarIndMapCreation.indeterminateDrawable.colorFilter = colorFilter
        progressBarIndUnzip.indeterminateDrawable.colorFilter = colorFilter
    }
}
